### DISCUSSION TITLE
Show stack byte-by-byte

### DISCUSSION THREAD

**Discussion by Occhioverde** [pwndbg/pwndbg#1753] [Q&A] [answered]
Good morning.

I'm currently using pwndbg to debug an Assembly program requested by my professor.

Since I'm working directly with the stack it would be nice if it was possible to show the entries byte-by-byte, but the default "screen" 
shows the addresses grouped by four (so each row is 32 bits), if I'm reading it correctly:

![pwndbg default stack view](https://github.com/pwndbg/pwndbg/assets/18429412/bc49f7db-e308-44b3-8948-02f557226b18)

Is there an option to disable this grouping and show an address for each line?

Thank you in advance for the help.

**Comment by disconnect3d** [comment, marked as answer]
Hi,

> Is there an option to disable this grouping and show an address for each line?


Not sure I understand this. What output would you want to see exactly? Something like the output of `hexdump $esp`?

Edit: Currently, the stack view shows ptr-sized words and if such a word is a valid pointer, we dereference it further to enhance the display to show the data under that pointer.

EDIT2: Here's how hexdump looks like:
![image](https://github.com/pwndbg/pwndbg/assets/10009354/c61e2028-a4f5-406d-b783-47dc2abbd81f)

**Reply by Occhioverde** [reply]
Thank you for the super quick reply!

If no other options are available, `hexdump` seems good to get what I want to see. The view I was trying to describe, however, is one equal to the stack one but with 1 byte words; something like:

```
00:0000│ esp 0xffff8ad0 ◂— 0x1
00:0001│     0xffff8ad1 ◂— 0x0
00:0002│     0xffff8ad2 ◂— 0x0
00:0003│     0xffff8ad3 ◂— 0x0
```

**Reply by disconnect3d** [reply]
Why would you need it? I mean, that is probably inefficient in terms of the space vs information provided, so I don't think many people would want to use that.

I would have to double check with `config` and `theme` (you can see those for yourself!) but I doubt we have another way to display stack in the context. 

One can change the color of stack (`set memory-stack-color <value>`), the number of lines it displays (`set context-stack-lines <N>`) and the number of pointers that will be dereferenced in a chain (`set dereference-limit <N>`, defaults to 5) but yeah, we don't have an alternative stack display.

**Reply by disconnect3d** [reply]
Given all this, I think it may make sense to have it configurable how the stack is displayed. I think displaying stack as  hexdump view may be useful for many, and could either be "another optional context view" (and then set e.g. with `set context-sections regs disasm stack hexstack`) or we could have a separate config that would tell the context how to render the stack, similarly as we have `set show-compact-regs on|off`.

**Reply by Occhioverde** [reply]
> I mean, that is probably inefficient in terms of the space vs information provided

Yea, I know. I just would find it more convenient to better visualize the stack management while writing and debugging my program, as the course I'm taking is an introduction to Assembly.

> but yeah, we don't have an alternative stack display

No problem, I think that `hexdump` will do just fine for my use case.

Thank you for your help!

**Comment by disconnect3d** [comment]
Closing this as `hexdump` seems to be the solution :)
